File: 이상욱/BOJ/m4w9/BOJ15681.java
package m4w9;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class BOJ15681 {

	static ArrayList<Integer>[] list, tree;
	static int p[], size[];

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int n = Integer.parseInt(st.nextToken());
		int r = Integer.parseInt(st.nextToken());
		int q = Integer.parseInt(st.nextToken());
		
		p = new int[n + 1];
		size = new int[n + 1];
		list = new ArrayList[n + 1];
		tree = new ArrayList[n + 1];
		
		for (int i = 0; i < list.length; i++) {
			list[i] = new ArrayList<>();
			tree[i] = new ArrayList<>();
		}
		
		
		for (int i = 1; i < n; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			list[u].add(v);
			list[v].add(u);
		}
		
		makeTree(r, -1);
		countSubtreeNodes(r);
		
		
		StringBuffer sb = new StringBuffer();
		while (q-- > 0) {
			int query = Integer.parseInt(br.readLine());
			sb.append(size[query]).append("\n");
		}
		
		bw.write(sb.toString());
		bw.flush();
		br.close();
		bw.close();
	}

	public static void makeTree(int currentNode, int parent) {
		
		for (int node : list[currentNode]) {
			
			if (node != parent) {
				tree[currentNode].add(node);
				p[node] = currentNode;
				makeTree(node, currentNode);
			}
		}
	}

	
	public static void countSubtreeNodes(int currentNode) {
		
		size[currentNode] = 1;
		
		for (int node : tree[currentNode]) {
			countSubtreeNodes(node);
			size[currentNode] += size[node];
		}
	}

}

File: 이상욱/BOJ/m4w9/BOJ15681.kt
package m4w9

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import kotlin.collections.ArrayList

lateinit var list : Array<ArrayList<Int>>
lateinit var tree : Array<ArrayList<Int>>
lateinit var p : IntArray
lateinit var size : IntArray
fun main() {

    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    var st = StringTokenizer(br.readLine())

    val n = st.nextToken().toInt()
    val r = st.nextToken().toInt()
    var q = st.nextToken().toInt()

    p = IntArray(n+1)
    size = IntArray(n+1)
    list = Array(n+1){ arrayListOf()}
    tree = Array(n+1){ arrayListOf()}

    for(i in 1 until n){
        st = StringTokenizer(br.readLine())

        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()

        list[u].add(v)
        list[v].add(u)
    }

    makeTree(r,-1)
    countSubTreeNodes(r)
    val sb = StringBuilder()

    while(q > 0){
        val query = br.readLine().toInt()

        sb.append(size[query]).append("\n")
        q--
    }

    bw.write(sb.toString())
    bw.flush()
    br.close()
    bw.close()

}

fun makeTree(currentNode : Int, parent : Int){
    list[currentNode].forEach {
        if( it != parent){
            tree[currentNode].add(it)
            p[it] = currentNode
            makeTree(it,currentNode)

        }
    }
}

fun countSubTreeNodes(currentNode: Int){
    size[currentNode] = 1

    tree[currentNode].forEach {
        countSubTreeNodes(it)
        size[currentNode] += size[it]
    }
}

